### PR TITLE
Feature vector

### PR DESCRIPTION
**Pull request by sccolbert**
Still needs tests.

### CODE REVIEW

**Comment by sccolbert** [comment]
Tests are added. This is mostly ready to go. Before merging, I want to discuss our policy on index validity.

File: src/collections/vector.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+   * #### Complexity
+   * Linear.
+   *
+   * #### Iterator Validity
+   * No changes.
+   *
+   * #### Undefined Behavior
+   * An `index` which is non-integral or out of range.
+   */
+  insert(index: number, value: T): void {
```
Comment (afshin) [inline src/collections/vector.ts]: Why not `this._array.splice(index, 0, value)`? Is there a performance consideration in using `splice`?

File: src/collections/vector.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+   * #### Complexity
+   * Linear.
+   *
+   * #### Iterator Validity
+   * Iterators pointing at the removed value and beyond are invalidated.
+   *
+   * #### Undefined Behavior
+   * An `index` which is non-integral or out of range.
+   */
+  remove(index: number): void {
```
Comment (afshin) [inline src/collections/vector.ts]: Same question here:
```
this._array.splice(index, 1);
```

File: src/collections/vector.ts (hunk 1 above)
Comment (sccolbert) [inline src/collections/vector.ts]: Splice always allocates a return value array, even if you don't care about it. It's also slower than this.

File: src/collections/vector.ts (hunk 2 above)
Comment (sccolbert) [inline src/collections/vector.ts]: Ditto.

**Pull request merged**